src/software/ai/intent/intent.h

```diff
@@ -90,17 +115,59 @@ class Intent
      */
     std::set<MotionConstraint> getMotionConstraints(void) const;
 
+    /**
+     * Gets the navigator params if this intent requires navigation. For example,
+     * StopIntent does not need to navigate, while MoveIntent does.
+     *
+     * @return navigator params
+     */
+    std::optional<NavigatorParams> getNavigatorParams() const;
+
+    /**
+     * Get PrimitiveMsg that underlies this Intent
+     *
+     * @return PrimitiveMsg
+     */
+    PrimitiveMsg getPrimitiveMsg() const;
+
+    /**
+     * Gets a PrimitiveMsg updated with destination and final_speed
+     * NOTE: only Intents that navigate update PrimitiveMsg
+     *
+     * @param destination The destination
+     * @param final_speed The final speed
+     *
+     * @return PrimitiveMsg updated with destination and final_speed
+     */
+    virtual PrimitiveMsg getUpdatedPrimitiveMsg(Point destination,
```

Comment:
I'm a little confused about how we're using this function. Could you explain at a high level why it's needed?

Reply:
Example: MoveIntent starts with a PrimitiveMsg to go to point C, starting from point A. Navigator finds a path that goes from A to B to C to avoid an obstacle. As MovePrimitives are currently, this means that the robot should go to point B and MovePrimitiveMsg needs to be changed to go to point B. 

getUpdatedPrimitiveMsg takes the new destination and final speed derived from the path, updates the PrimitiveMsg and returns the updated PrimitiveMsg

Reply:
This feels a bit hacky, like we're moving some of the navigator work into each primitive. 

Based on this, why not keep some sort of Intent->PrimitiveMsg visitor that takes this information + an intent? This would allow decoupling of intents from primitives entirely, which seems like a better option from a architecture perspective.

Reply:
A visitor would abstract away PrimitiveMsg, but it would require each Intent to store all the fields necessary to create a PrimitiveMsg, which is quite a bit of duplication. To add a new primitive, in addition to implementing in firmware and defining in protobuf, one would have to define it again as an Intent. 

I'll poke around with visitor on PrimitiveMsg to update them with the Path. I think that would be simplest